### PR TITLE
[SPARK-21608][SPARK-9221][SQL] Window rangeBetween() API should allow literal boundary

### PR DESCRIPTION
**Pull request by jiangxb1987**
## What changes were proposed in this pull request?

Window rangeBetween() API should allow literal boundary, that means, the window range frame can calculate frame of double/date/timestamp.

Example of the use case can be:
```
SELECT
	val_timestamp,
	cate,
	avg(val_timestamp) OVER(PARTITION BY cate ORDER BY val_timestamp RANGE BETWEEN CURRENT ROW AND interval 23 days 4 hours FOLLOWING)
FROM testData
```

This PR refactors the Window `rangeBetween` and `rowsBetween` API, while the legacy user code should still be valid.

## How was this patch tested?

Add new test cases both in `DataFrameWindowFunctionsSuite` and in `window.sql`.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala [hunk 1]
```diff
@@ -84,7 +84,7 @@ object Window {
    *
    * @since 2.1.0
    */
-  def unboundedPreceding: Long = Long.MinValue
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: We are unable to change the public interface.

**Comment by JoshRosen** [comment]
(test comment to test PR dashboard linking)

**Comment by JoshRosen** [comment]
(test comment to test PR dashboard linking)
--

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 2]
```diff
@@ -821,9 +821,12 @@ object TypeCoercion {
     }
 
     private def createBoundaryCast(boundary: Expression, dt: DataType): Expression = {
-      boundary match {
-        case e: SpecialFrameBoundary => e
-        case e: Expression if e.dataType != dt && Cast.canCast(e.dataType, dt) => Cast(e, dt)
+      (boundary, dt) match {
+        case (e: SpecialFrameBoundary, _) => e
+        case (e: Expression, _: DateType) => e
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: nit: just `e` instead of `e: Expression`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 3]
```diff
@@ -821,9 +821,12 @@ object TypeCoercion {
...
+        case (e: Expression, _: TimestampType) => e
+        case (e: Expression, t: DataType) if e.dataType != t && Cast.canCast(e.dataType, t) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: nit: just `t` instead of `t: DataType`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala [hunk 4]
```diff
@@ -89,7 +89,11 @@ case class WindowSpecDefinition(
     elements.mkString("(", " ", ")")
   }
 
-  private def isValidFrameType(ft: DataType): Boolean = orderSpec.head.dataType == ft
+  private def isValidFrameType(ft: DataType): Boolean = (orderSpec.head.dataType, ft) match {
+    case (DateType, IntegerType) => true
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: don't we need `CalendarIntervalType` for `DateType` too?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowExec.scala [hunk 5]
```diff
@@ -139,7 +141,12 @@ case class WindowExec(
         }
 
         // Create the projection which returns the current 'value' modified by adding the offset.
-        val boundExpr = Add(expr, Cast(boundOffset, expr.dataType))
+        val boundExpr = (expr.dataType, boundOffset.dataType) match {
+          case (DateType, IntegerType) => DateAdd(expr, boundOffset)
+          case (TimestampType, CalendarIntervalType) =>
+            TimeAdd(expr, boundOffset, Some(DateTimeUtils.defaultTimeZone.getID))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowExec.scala]: shall we use session timezone here?

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala [hunk 6]
```diff
@@ -159,6 +159,17 @@ object Window {
   }
 
   /**
+   * @param start boundary start, inclusive. The frame is unbounded if this evals to
+   *              the minimum long value (`Window.unboundedPreceding`).
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: `.. the minimum long value ...` this is not true now, we use a special value `functions.unboundedPreceding`

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala [hunk 7]
```diff
@@ -159,6 +159,17 @@ object Window {
...
+   * @param end boundary end, inclusive. The frame is unbounded if this evals to the
+   *            maximum long value (`Window.unboundedFollowing`).
+   * @since 2.3.0
+   */
+  def rowsBetween(start: Column, end: Column): WindowSpec = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: hmmm, how useful is this API? `rowsBetween` only accept integer and max/min long, and it's ok if we only have `rowsBetween(start: Long, end: Long)`

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 8]
```diff
@@ -777,6 +778,29 @@ object functions {
   //////////////////////////////////////////////////////////////////////////////////////////////
   // Window functions
   //////////////////////////////////////////////////////////////////////////////////////////////
+  /**
+   * Window function: returns the value that represents the first row in the window partition.
+   *
+   * @group window_funcs
+   * @since 2.3.0
+   */
+  def unboundedPreceding(): Column = lit(Window.unboundedPreceding)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: can't we just return `Column(UnboundedPreceding)`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 3 above)
Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: `Cast.canCast()` only accepts `DataType` as input param.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala (hunk 4 above)
Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: I think `CalendarIntervalType` is only used with `TimestampType`, cc @hvanhovell @ueshin

Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: I think `CalendarIntervalType` is only used with `TimestampType`, cc @hvanhovell @ueshin

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 9]
```diff
@@ -821,9 +821,12 @@ object TypeCoercion {
...
 
     private def createBoundaryCast(boundary: Expression, dt: DataType): Expression = {
-      boundary match {
-        case e: SpecialFrameBoundary => e
-        case e: Expression if e.dataType != dt && Cast.canCast(e.dataType, dt) => Cast(e, dt)
+      (boundary, dt) match {
+        case (e: SpecialFrameBoundary, _) => e
+        case (e, _: DateType) => e
+        case (e, _: TimestampType) => e
+        case (e: Expression, t: DataType) if e.dataType != t && Cast.canCast(e.dataType, t) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: `t` is always `DataType`, so we don't need to write `t: DataType`

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala [hunk 10]
```diff
@@ -167,17 +167,17 @@ object Window {
    * current row.
    *
    * We recommend users use `Window.unboundedPreceding`, `Window.unboundedFollowing`,
-   * and `Window.currentRow` to specify special boundary values, rather than using integral
+   * and `Window.currentRow` to specify special boundary values, rather than using literal
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: long values?

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala [hunk 11]
```diff
@@ -210,6 +210,17 @@ object Window {
     spec.rangeBetween(start, end)
   }
 
+  /**
+   * @param start boundary start, inclusive. The frame is unbounded if the expression is
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: we should copy-paste the document from `rangeBetween(start: Long, end: Long)` to here, and replace `long values` with `literal values`. We should also update the examle to show how to use special range boundary.

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala [hunk 12]
```diff
@@ -210,6 +210,20 @@ class WindowSpec private[sql](
   }
 
   /**
+   * @param start boundary start, inclusive. The frame is unbounded if the expression is
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala [hunk 13]
```diff
@@ -167,17 +167,17 @@ object Window {
...
    *
    * A range-based boundary is based on the actual value of the ORDER BY
    * expression(s). An offset is used to alter the value of the ORDER BY expression, for
    * instance if the current order by expression has a value of 10 and the lower bound offset
    * is -3, the resulting lower bound for the current row will be 10 - 3 = 7. This however puts a
    * number of constraints on the ORDER BY expressions: there can be only one expression and this
-   * expression must have a numerical data type. An exception can be made when the offset is 0,
-   * because no value modification is needed, in this case multiple and non-numeric ORDER BY
-   * expression are allowed.
+   * expression must have a literal data type. An exception can be made when the offset is
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: literal -> numerical or date or timestamp data type

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala [hunk 14]
```diff
@@ -210,6 +210,57 @@ object Window {
...
 
+  /**
+   * Creates a [[WindowSpec]] with the frame boundaries defined,
+   * from `start` (inclusive) to `end` (inclusive).
+   *
+   * Both `start` and `end` are relative to the current row. For example, "lit(0)" means
+   * "current row", while "lit(-1)" means one off before the current row, and "lit(5)" means the
+   * five off after the current row.
+   *
+   * Users should use `unboundedPreceding()`, `unboundedFollowing()`, and `currentRow()` to specify
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: let's mention that these 3 functions are from `object functions`

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala [hunk 15]
```diff
@@ -146,22 +146,22 @@ class WindowSpec private[sql](
   /**
    * Defines the frame boundaries, from `start` (inclusive) to `end` (inclusive).
    *
-   * Both `start` and `end` are relative from the current row. For example, "0" means "current row",
-   * while "-1" means one off before the current row, and "5" means the five off after the
-   * current row.
+   * Both `start` and `end` are relative from the current row. For example, "0" means
+   * "current row", while "-1" means one off before the current row, and "5" means the five off
+   * after the current row.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala]: why this change? seems you only change the line length?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala (hunk 4 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: can we check with other databases? seems `RANGE BETWEEN CURRENT ROW AND interval 2 days FOLLOWING` is reasonable for date type column.

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala (hunk 15 above)
Comment (jiangxb1987) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala]: The original line length exceeds the 100 charactors limit...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala (hunk 4 above)
Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: I checked and HIVE support that, so let's also allow this.

Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: Yea, we should allow this, but this is a bit corner case and not very straight-forward to implement, maybe we can leave this as a follow-up?

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala [hunk 16]
```diff
@@ -167,17 +167,17 @@ object Window {
...
    * A range-based boundary is based on the actual value of the ORDER BY
    * expression(s). An offset is used to alter the value of the ORDER BY expression, for
    * instance if the current order by expression has a value of 10 and the lower bound offset
    * is -3, the resulting lower bound for the current row will be 10 - 3 = 7. This however puts a
    * number of constraints on the ORDER BY expressions: there can be only one expression and this
-   * expression must have a numerical data type. An exception can be made when the offset is 0,
-   * because no value modification is needed, in this case multiple and non-numeric ORDER BY
-   * expression are allowed.
+   * expression must have a numerical data type. An exception can be made when the offset is
+   * unbounded, because no value modification is needed, in this case multiple and non-literal
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: `non-literal` -> `non-numeric`

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala [hunk 17]
```diff
@@ -210,6 +210,57 @@ object Window {
...
+   * transformed to [[org.apache.spark.sql.catalyst.expressions.SpecialFrameBoundary]]s.
+   *
+   * A range-based boundary is based on the actual value of the ORDER BY
+   * expression(s). An offset is used to alter the value of the ORDER BY expression, for
+   * instance if the current order by expression has a value of 10 and the lower bound offset
+   * is -3, the resulting lower bound for the current row will be 10 - 3 = 7. This however puts a
+   * number of constraints on the ORDER BY expressions: there can be only one expression and this
+   * expression must have a numerical/date/timestamp data type. An exception can be made when the
+   * offset is unbounded, because no value modification is needed, in this case multiple and
+   * non-literal ORDER BY expression are allowed.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: non-numerical/date/timestamp data type

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala [hunk 18]
```diff
@@ -146,22 +146,22 @@ class WindowSpec private[sql](
...
    * A range-based boundary is based on the actual value of the ORDER BY
    * expression(s). An offset is used to alter the value of the ORDER BY expression, for
    * instance if the current order by expression has a value of 10 and the lower bound offset
    * is -3, the resulting lower bound for the current row will be 10 - 3 = 7. This however puts a
    * number of constraints on the ORDER BY expressions: there can be only one expression and this
-   * expression must have a numerical data type. An exception can be made when the offset is 0,
-   * because no value modification is needed, in this case multiple and non-numeric ORDER BY
-   * expression are allowed.
+   * expression must have a numerical data type. An exception can be made when the offset is
+   * unbounded, because no value modification is needed, in this case multiple and non-literal
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala [hunk 19]
```diff
@@ -210,6 +210,59 @@ class WindowSpec private[sql](
...
+   * transformed to [[org.apache.spark.sql.catalyst.expressions.SpecialFrameBoundary]]s.
+   *
+   * A range-based boundary is based on the actual value of the ORDER BY
+   * expression(s). An offset is used to alter the value of the ORDER BY expression, for
+   * instance if the current order by expression has a value of 10 and the lower bound offset
+   * is -3, the resulting lower bound for the current row will be 10 - 3 = 7. This however puts a
+   * number of constraints on the ORDER BY expressions: there can be only one expression and this
+   * expression must have a numerical/date/timestamp data type. An exception can be made when the
+   * offset is unbounded, because no value modification is needed, in this case multiple and
+   * non-literal ORDER BY expression are allowed.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala]: ditto

**Comment by cloud-fan** [comment]
thanks, merging to master!
